driver-core/src/main/com/mongodb/MongoSocketException.java

```diff
@@ -33,7 +33,7 @@ public class MongoSocketException extends MongoException {
      * @param msg the message
      * @param e the cause
      */
-    MongoSocketException(final String msg, final ServerAddress serverAddress, final Throwable e) {
+    public MongoSocketException(final String msg, final ServerAddress serverAddress, final Throwable e) {
```

Comment:
We generally need them to be public for the reason mentioned.  The `VisibleForTesting` annotation is not really accurate, but I suppose we could use it for constructors that we don't otherwise want to be public.

Reply:
> We generally need them to be public for the reason mentioned.

Being `public` and being part of the public API is not the same. In this case, the constructor has to be `public`, but, as far as I can judge from this conversation, does not have to be part of the public API.

> The `VisibleForTesting` annotation is not really accurate

I agree, it should not be used here. But we can document the constructor as not being part of the public API, similarly to how the constructor of `MongoConnectionPoolClearedException` is documented. By doing so, we reduce the API surface. Do you see reasons why this is a bad idea?

Reply:
There's just a lack of consistency in that the large majority of our exception class constructors are already part of the API and there is not a very clear path to changing that (would we deprecate them all?).  So making this one not part of the API doesn't help us all that much.  I wouldn't be against a ticket that considers options for shrinking out total constructor API.

Reply:
Created https://jira.mongodb.org/browse/JAVA-4942 and https://jira.mongodb.org/browse/JAVA-4943.